node/node.go

```diff
@@ -1285,6 +1285,36 @@ func (app *App) setupDBs(ctx context.Context, lg log.Log, dbPath string) error {
 
 // Start starts the Spacemesh node and initializes all relevant services according to command line arguments provided.
 func (app *App) Start(ctx context.Context) error {
+	err := app.startSynchronous(ctx)
+	if err != nil {
+		return err
+	}
+	defer events.ReportError(events.NodeError{
+		Msg:   "node is shutting down",
+		Level: zapcore.InfoLevel,
+	})
+	// TODO: pass app.eg to components and wait for them collectively
+	if app.ptimesync != nil {
+		app.eg.Go(func() error {
+			app.errCh <- app.ptimesync.Wait()
+			return nil
+		})
+	}
+	// app blocks until it receives a signal to exit
+	// this signal may come from the node or from sig-abort (ctrl-c)
+	select {
+	case <-ctx.Done():
+		return nil
+	case err = <-app.errCh:
+		return err
+	}
+}
+
+func (app *App) startSynchronous(ctx context.Context) error {
+	// notify anyone who might be listening that the app has finished starting.
+	// this can be used by, e.g., app tests.
+	defer close(app.started)
```

Comment:
Now it will close the channel even in the case of an error during a start. Should it be that way?

Reply:
That is the intention, consider this code:

```go
g.Go(func() error {
	return app1.Start(ctx)
})
<-app1.Started()
```

If there's an error at startup and the started channel is not closed, the caller gets stuck, it doesn't matter that the error is returned through the error group, because the code is not able to progress to the point where it waits on the error group.

Reply:
OK, now I understand what you want to do in this PR, actually. But TBH, I don't understand the purpose of the `started` channel. It now synchronizes to the point in time where `startSynchronous()` quits, but how is it useful?

Reply:
Do we even need the channel any more? It is closed when `Start` returns. So if `Start` returns `nil` the start was successful, otherwise it wasn't.

Reply:
Hey @poszu & @fasmat the thing is that `Start` runs the app,  so if all goes well `Start` won't return until the app is shutting down. I believe the purpose of the started channel is to let the caller know when it is safe to call the APIs exposed by the app. An alternative would be to separate all listening from serving operations and then app startup could be done in two stages, first setting up all services to listen and a second call to start serving those services. However I think that would be a much larger change.

Reply:
Yeah, I momentarily forgot that `app.Start` is a blocking call. Forget my comment about the channel not being needed any more 🙂

Reply:
@fasmat Hehe, easy to do, I have done the same :D

@poszu You're right, its a bit of a kludge, you don't actually know for sure that the services are going to be listening after receiving on the channel. However if the channel was removed all you could do is wait. So I think the channel use is preferable to that, or maybe I've misunderstood the original purpose of this channel, but I can't think what else it's for.

Reply:
@poszu I added a ticket for the 2 stage startup. - https://github.com/spacemeshos/go-spacemesh/issues/4892